lib/datahub-client/data_platform_catalogue/client/search.py

```diff
@@ -107,35 +113,28 @@ def search(
     def _parse_search_results(self, response) -> Tuple[list, list]:
         page_results = []
         malformed_result_urns = []
+
+        EntityMapping = namedtuple('EntityMapping', ['result_type', 'parsing_function'])
+
+        entity_types_map = {
+            DATASET: EntityMapping(ResultType.TABLE, self._parse_dataset),
+            CHART: EntityMapping(ResultType.CHART, self._parse_dataset),
+            CONTAINER: EntityMapping(ResultType.DATABASE, self._parse_container),
+            DASHBOARD: EntityMapping(ResultType.DASHBOARD, self._parse_container)
+        }
+
         for result in response["searchResults"]:
             entity = result["entity"]
             entity_type = entity["type"]
             entity_urn = entity["urn"]
             matched_fields = self._get_matched_fields(result=result)
 
             try:
-                if entity_type == "DATASET":
-                    parsed_result = self._parse_dataset(
-                        entity, matched_fields, ResultType.TABLE
-                    )
-                    page_results.append(parsed_result)
-                elif entity_type == "CHART":
-                    parsed_result = self._parse_dataset(
-                        entity, matched_fields, ResultType.CHART
-                    )
-                    page_results.append(parsed_result)
-                elif entity_type == "CONTAINER":
-                    parsed_result = self._parse_container(
-                        entity, matched_fields, ResultType.DATABASE
-                    )
-                    page_results.append(parsed_result)
-                elif entity_type == "DASHBOARD":
-                    parsed_result = self._parse_container(
-                        entity, matched_fields, ResultType.DASHBOARD
-                    )
-                    page_results.append(parsed_result)
-                else:
-                    raise Exception
+                parsed_result = entity_types_map[entity_type].parsing_function(entity, matched_fields, entity_types_map[entity_type].result_type)
```

Comment:
Since there's a lot going on in this search class, I'm wondering if it's worth going a step further and extracting out helper classes for parsing the various entities.

something like

```python
class Parser:
  def parse_types_and_subtypes(self): ...

class ContainerParser(Parser):
  def parse(self, entity, matched_fields, result_type): ...
  
class DatasetParser(Parser):
  def parse(self, entity, matched_fields, result_type): ...
```

In that case

- `_parse_dataset`, `_parse_container` and `_parse_types_and_subtypes` become public methods of the new classes, and can be unit tested
- instead of a parsing function, you map to an instance of one of these classes, and call its `parse` method

what do you think?

Reply:
This is a nice idea! We've moved all these changes to the larger refactor PR where the mapping is a bit more complicated, perhaps we could work this in